text/1028-cli-hooks.rst

```diff
@@ -56,70 +69,182 @@ are always executed in WSL.
+order.
+
+
+Watch Configuration
+===================
+
+Sometimes in a Gel project there's a need to respond to some file changes: run
+a migration due to schema change or run code generators due to query file
+change. The CLI already supports watching the schema for changes and applying
+them, but we can generalize this mechanism and make it more flexible, much
+like the hooks that respond to CLI commands.
+
+The idea is to have a mechanism for specifying a file system path you want to
+watch for modifications and a script that gets triggered by those changes.
+Unlike hooks, there are no "before" and "after" triggers here. Only one kind
+of response is possible: trigger when a change to the watched entity is
+detected.
+
+The watch configuration will be described in the ``gel.toml`` project manifest
+file in the ``[watch]`` table. We will introduce a ``[watch.files]`` sub-table
+for watching file system changes and triggering scripts.
+
+The general structure of watch tables is going to use the key (potentially
+quoted) to specify *what is being watched* and the corresponding value to
+specify the script that will be triggered by changes.
+
+The ``gel watch`` command (without needing further options) would then be used
+to start the watch process and monitor whatever is specified in the
+``gel.toml``. By default ``gel watch`` will only watch the files specified in
+the ``gel.toml`` config and execute the trigger scripts. Running ``gel watch
+--migrate`` will additionally monitor schema changes and perform real-time
+migrations.
+
+This is a backwards incompatible change compared to how ``edgedb watch`` operates
+now.
+
+The output of the scripts will then appear in the same TTY as the ``gel
+watch`` command.
+
+Multiple consecutive file system change events will be debounced into a single
+event. For example, given scripts ``src/**`` and ``*.py`` and writes to
+``src/a.py``, ``src/b.py`` and ``src/c.rs``, both scripts will trigger exactly
+once. To achive this, ``gel watch``` will wait for a small amount of time
+(100ms) after receiving first event before executing matching scripts. This is
+mostly to reduce unnecessary multiple triggers for the same watched entity.
+
+Another consequence of executing triggered scripts in the background is that
+sometimes the changes are invalid in some way and the script will fail. This
+is considered part of normal operation (e.g. syntax error in a saved query
+file) and the scripts should be such that failure does not create some
+non-recoverable state.
+
+Due to the nature of monitoring changes, we cannot guarantee any particular
+order in which the watch scripts will be triggered when multiple changes occur
+at once. This is independent of the source of multiple triggers: whether it is
+due to multiple files being updated or multiple watch rules matching the same
+file. The scripts are effectively triggered asynchronously and the exact order
+is an implementation detail that should not be relied upon.
```

Comment:
This differs from current implementation:
```python
changed_paths = wait_for_changes()

for matcher in matchers:
  is_matched = any(matcher.matches(p) for p changed_paths)
  if is_matched:
    await run_script(matcher.script)
```

Differences:
- the order of fs events does not matter,
- order of matchers (and executed scripts) is fixed between runs, but it is an implementation detail,
- scripts are run sequentially (I wait for first to complete before running the next one).

Is this behavior ok? It can be changed if we have a good reason to do so.

Reply:
@vpetrovykh